pydantic_evals/pydantic_evals/datasets.py

```diff
@@ -0,0 +1,376 @@
+        cls._assessment_registry = cls._assessment_registry.copy()
+
+    @classmethod
+    def assessment_registry(cls) -> dict[str, AssessmentFunction[InputsT, OutputT, MetadataT]]:
+        combined_registry: dict[str, AssessmentFunction[Any, Any, Any]] = {**get_default_registry()}
+        for c in cls.__mro__[::-1]:
+            if hasattr(c, '_assessment_registry'):
+                combined_registry.update(c._assessment_registry)  # type: ignore
+        return combined_registry
+
+    @classmethod
+    @functools.cache
+    def _row_type(cls) -> type[EvaluationRow[InputsT, OutputT, MetadataT]]:
+        return EvaluationRow[cls._params]  # type: ignore
+
+    @classmethod
+    @functools.cache
+    def _params(cls) -> tuple[type[InputsT], type[OutputT], type[MetadataT]]:
+        for c in cls.__mro__:
+            metadata = getattr(c, '__pydantic_generic_metadata__')
+            if len(args := metadata.get('args', ())) == 3:
+                return args
+            elif len(args := getattr(c, '__args__', ())) == 3:
+                return args
+        raise ValueError(f'Could not determine the generic parameters for {cls}')
+
+    @classmethod
+    def assessment(
+        cls, f: AssessmentFunction[InputsT, OutputT, MetadataT]
+    ) -> AssessmentFunction[InputsT, OutputT, MetadataT]:
+        """Decorator that registers an assessment function in the class-specific registry.
+
+        This provides a generic-type-checking alternative to the `@assessment` decorator.
+        """
+        cls._assessment_registry[f.__name__] = f
+        return f
+
+    def eval_cases(self) -> list[EvaluationRow[InputsT, OutputT, MetadataT]]:
+        registry = self.assessment_registry()
+
+        evaluation_rows: list[EvaluationRow[InputsT, OutputT, MetadataT]] = []
+        errors: list[ValueError] = []
+        row_type = self._row_type()
+        for row in self.rows:
+            assessments: list[Assessment[Any, Any, Any]] = []
+            for spec in row.assessments + self.default_assessments:
+                try:
+                    assessment = Assessment[InputsT, OutputT, MetadataT].from_registry(registry, row.name, spec)
+                except ValueError as e:
+                    errors.append(e)
+                    continue
+                assessments.append(assessment)
+            evaluation_rows.append(
+                row_type(
+                    name=row.name,
+                    inputs=row.inputs,
+                    metadata=row.metadata,
+                    expected_output=row.expected_output,
+                    assessments=assessments,
+                )
+            )
+        if errors:
+            raise ExceptionGroup('Failed to load assessments from registry', errors)
+        return evaluation_rows
+
+    # TODO: Task: Always save a schema file when saving the dataset
+    def save(self, path: Path | str = DEFAULT_DATASET_PATH, schema_ref: str | None = None) -> None:
```

Comment:
This isn't used anywhere. How does it differ from the stuff in generate_dataset_files?

Reply:
ooooh, it took me a long time to see that generate_dataset_files generates files with no rows. that needs a better name. is that useful? isn't writing a single case by hand quite tricky?

Reply:
maybe rename to `to_yaml`?

Reply:
yeah probably, I think this needs a bit of massaging